src/storage/routing/RegexRouterRule.ts

```diff
@@ -0,0 +1,56 @@
+import type { Representation } from '../../ldp/representation/Representation';
+import type { ResourceIdentifier } from '../../ldp/representation/ResourceIdentifier';
+import { UnsupportedHttpError } from '../../util/errors/UnsupportedHttpError';
+import { trimTrailingSlashes } from '../../util/Util';
+import type { ResourceStore } from '../ResourceStore';
+import { RouterRule } from './RouterRule';
+
+interface RegexEntry {
+  regex: RegExp;
+  store: ResourceStore;
+}
+
+/**
+ * Routes requests to a store based on the path of the identifier.
+ * The identifier will be stripped of the base URI after which regexes will be used to find the correct store.
+ * The trailing slash of the base URI will still be present so the first character a regex can match would be that `/`.
+ * This way regexes such as `/container/` can match containers in any position.
+ *
+ * In case none of the regexes match an error will be thrown.
+ */
+export class RegexRouterRule extends RouterRule {
+  private readonly base: string;
+  private readonly regexes: RegexEntry[];
+
+  /**
+   * The keys of the `storeMap` will be converted into actual RegExp objects that will be used for testing.
+   */
+  public constructor(base: string, storeMap: { [regex: string]: ResourceStore }) {
+    super();
+    this.base = trimTrailingSlashes(base);
+    this.regexes = Object.keys(storeMap).map((regex): RegexEntry =>
+      ({ regex: new RegExp(regex, 'u'), store: storeMap[regex] }));
```

Comment:
Problem is that we then won't be able to configure this class in Components.js, which is needed for this one.

Reply:
I get that; so I'd still have the JSON-configurable version as a fallback, but also make sure that we can properly construct if we want to.

So:

```
public constructor(base: string, storeMap: Map<RegExp, String> | { [regex: string]: ResourceStore }) {
...
if (!isMap(storeMap)))
  storeMap = toRegexMap(storeMap);
```

Reply:
Components.js generator doesn't like this though, since it can't convert the first part

Reply:
Ah ok, but that should be filed as a Components.js issue then, and marked as a todo the code (or as an issue on this repo).

We can't have Components.js influence our design; and if it does, we should annotate.

Reply:
This might get quite ugly though. The union type is already quite verbose, and I don't think there's a way for components.js to directly support this so it might need special annotations not helping the situation. But I'll leave that comment up to @rubensworks

Reply:
OK, thanks; just document this as an issue/todo somewhere and we can move on for now.